File: internal/checks/target_blank_noopener.go
package checks

import (
	"bytes"
	"context"
	"fmt"
	"net/url"
	"sort"
	"strings"

	"golang.org/x/net/html"

	"github.com/londonmax12/hyperz/internal/httpclient"
	"github.com/londonmax12/hyperz/internal/page"
	"github.com/londonmax12/hyperz/internal/scope"
)

// TargetBlankNoopener flags <a target="_blank"> (and <area target="_blank">,
// <form target="_blank">) that do not carry rel="noopener" or rel="noreferrer".
//
// A link that opens a new browsing context hands the destination page a live
// window.opener reference back to the originating tab. Without noopener /
// noreferrer the destination can navigate the opener via
//
//	window.opener.location = "https://phish.example/"
//
// turning a one-click outbound link into reverse-tabnabbing: the user comes
// back to the original tab and finds a convincing phishing page in place of
// what they trusted.
//
// Modern browsers (Chrome 88+, Firefox 79+, Safari 12.1+) default anchors
// and area elements with target="_blank" to noopener, so this check is
// primarily defense-in-depth: older browsers, embedded webviews, and
// <form target="_blank"> (which did not get the same default treatment in
// every engine) all still rely on the explicit attribute. Severity is Low
// for same-origin destinations (the attacker would already control the
// origin) and Medium for cross-origin destinations, where the phishing
// surface is real.
type TargetBlankNoopener struct{}

func (TargetBlankNoopener) Name() string { return "target-blank-noopener" }

func (TargetBlankNoopener) Level() Level { return LevelPassive }

const targetBlankNoopenerBodyCap = 2 << 20

func (c TargetBlankNoopener) Run(ctx context.Context, client *httpclient.Client, _ *scope.Scope, p page.Page) ([]Finding, error) {
	snap, err := ensureResponse(ctx, client, p, targetBlankNoopenerBodyCap)
	if err != nil {
		return nil, err
	}
	if !isHTMLContentType(snap.Headers.Get("Content-Type")) {
		return nil, nil
	}
	if len(snap.Body) == 0 {
		return nil, nil
	}

	pageURL, err := url.Parse(p.URL)
	if err != nil || pageURL.Host == "" {
		return nil, nil
	}

	candidates := parseTargetBlankCandidates(snap.Body, pageURL)
	if len(candidates) == 0 {
		return nil, nil
	}

	var findings []Finding
	seen := map[string]struct{}{}

	for _, cand := range candidates {
		key := cand.tag + "|" + cand.resolved.String()
		dedupeKey := MakeKey(c.Name(), ScopePage, p.URL, "ref:"+key)
		if _, dup := seen[dedupeKey]; dup {
			continue
		}
		seen[dedupeKey] = struct{}{}

		crossOrigin := !strings.EqualFold(cand.resolved.Hostname(), pageURL.Hostname())
		severity := SeverityLow
		if crossOrigin {
			severity = SeverityMedium
		}

		findings = append(findings, Finding{
			Check:       c.Name(),
			Target:      p.URL,
			URL:         p.URL,
			Severity:    severity,
			Title:       buildTargetBlankTitle(cand.tag, crossOrigin),
			Detail:      buildTargetBlankDetail(p.URL, cand, crossOrigin),
			CWE:         "CWE-1022",
			OWASP:       "A05:2021 Security Misconfiguration",
			Remediation: buildTargetBlankRemediation(cand.tag),
			Evidence:    BuildEvidence("GET", p.URL, snap.Status, snap.Headers, ""),
			DedupeKey:   dedupeKey,
		})
	}

	sort.SliceStable(findings, func(i, j int) bool {
		return findings[i].DedupeKey < findings[j].DedupeKey
	})

	return findings, nil
}

// targetBlankCandidate is one (tag, href/action) pair the check should
// report. raw is the attribute as it appeared in the document, resolved is
// the absolute URL the browser would navigate to.
type targetBlankCandidate struct {
	tag      string
	raw      string
	resolved *url.URL
}

// parseTargetBlankCandidates walks body once and returns every <a>, <area>,
// or <form> with target="_blank" that lacks rel="noopener" / "noreferrer"
// and points at a network-scheme URL. baseURL is updated when a <base href>
// is observed so relative URLs resolve against the document base rather
// than the page URL when an explicit base is in play.
func parseTargetBlankCandidates(body []byte, pageURL *url.URL) []targetBlankCandidate {
	z := html.NewTokenizer(bytes.NewReader(body))

	baseURL := *pageURL
	baseURLPtr := &baseURL

	var out []targetBlankCandidate

	for {
		tt := z.Next()
		if tt == html.ErrorToken {
			break
		}
		if tt != html.StartTagToken && tt != html.SelfClosingTagToken {
			continue
		}
		tag, hasAttr := z.TagName()
		tagName := string(tag)

		switch tagName {
		case "base":
			if href := attrValue(z, hasAttr, "href"); href != "" {
				if u, err := url.Parse(strings.TrimSpace(href)); err == nil {
					baseURLPtr = baseURL.ResolveReference(u)
				}
			}
			continue
		case "a", "area", "form":
		default:
			continue
		}
		if !hasAttr {
			continue
		}

		hrefAttr := "href"
		if tagName == "form" {
			hrefAttr = "action"
		}

		var href, target, rel string
		var haveHref, haveTarget, haveRel bool
		for {
			key, val, more := z.TagAttr()
			switch strings.ToLower(string(key)) {
			case hrefAttr:
				if !haveHref {
					href = string(val)
					haveHref = true
				}
			case "target":
				if !haveTarget {
					target = string(val)
					haveTarget = true
				}
			case "rel":
				if !haveRel {
					rel = string(val)
					haveRel = true
				}
			}
			if !more {
				break
			}
		}

		if !strings.EqualFold(strings.TrimSpace(target), "_blank") {
			continue
		}
		if relHasNoopenerOrNoreferrer(rel) {
			continue
		}
		resolved, ok := resolveNoopenerHref(href, baseURLPtr)
		if !ok {
			continue
		}
		out = append(out, targetBlankCandidate{
			tag:      tagName,
			raw:      href,
			resolved: resolved,
		})
	}
	return out
}

// relHasNoopenerOrNoreferrer reports whether the (space-separated) rel
// attribute contains either token. Matching is case-insensitive per the
// HTML spec.
func relHasNoopenerOrNoreferrer(rel string) bool {
	for _, tok := range strings.Fields(rel) {
		switch strings.ToLower(tok) {
		case "noopener", "noreferrer":
			return true
		}
	}
	return false
}

// resolveNoopenerHref returns the absolute URL the browser would navigate
// to, or (nil, false) for non-network values (empty, javascript:, mailto:,
// tel:, data:, fragment-only). Resolution applies baseURL which may be the
// page URL itself, or a <base href> override.
func resolveNoopenerHref(raw string, baseURL *url.URL) (*url.URL, bool) {
	trimmed := strings.TrimSpace(raw)
	if trimmed == "" {
		return nil, false
	}
	lower := strings.ToLower(trimmed)
	if strings.HasPrefix(lower, "javascript:") ||
		strings.HasPrefix(lower, "mailto:") ||
		strings.HasPrefix(lower, "tel:") ||
		strings.HasPrefix(lower, "data:") ||
		strings.HasPrefix(lower, "#") {
		return nil, false
	}
	parsed, err := url.Parse(trimmed)
	if err != nil {
		return nil, false
	}
	resolved := baseURL.ResolveReference(parsed)
	if resolved.Host == "" {
		return nil, false
	}
	if !strings.EqualFold(resolved.Scheme, "http") && !strings.EqualFold(resolved.Scheme, "https") {
		return nil, false
	}
	return resolved, true
}

func buildTargetBlankTitle(tag string, crossOrigin bool) string {
	origin := "same-origin"
	if crossOrigin {
		origin = "cross-origin"
	}
	return fmt.Sprintf("<%s target=\"_blank\"> to %s URL without rel=\"noopener\"", tag, origin)
}

func buildTargetBlankDetail(pageURL string, cand targetBlankCandidate, crossOrigin bool) string {
	var b strings.Builder
	fmt.Fprintf(&b, "Page %s contains <%s target=\"_blank\" %s=%q> resolving to %s without rel=\"noopener\" or rel=\"noreferrer\". ",
		pageURL, cand.tag, attrLabelFor(cand.tag), cand.raw, cand.resolved.String())
	b.WriteString("The new browsing context receives a live window.opener handle to this page; ")
	b.WriteString("script in the destination can call window.opener.location = \"...\" to silently navigate this tab to a phishing page (reverse tabnabbing). ")
	if crossOrigin {
		b.WriteString("The destination is cross-origin, so any compromise or hostile content on that origin can pivot back into this site's tab.")
	} else {
		b.WriteString("The destination is same-origin, so direct impact is limited, but the missing attribute is still defense-in-depth worth fixing.")
	}
	return b.String()
}

func buildTargetBlankRemediation(tag string) string {
	switch tag {
	case "form":
		return "Add rel=\"noopener noreferrer\" to the <form> element. Forms with target=\"_blank\" did not get the same browser-default noopener treatment that anchors received, so the explicit attribute is the only portable guarantee."
	default:
		return "Add rel=\"noopener noreferrer\" to the element (e.g. <a href=\"...\" target=\"_blank\" rel=\"noopener noreferrer\">). Modern browsers default anchors with target=\"_blank\" to noopener, but older browsers, embedded webviews, and any code that opens windows via JavaScript still rely on the explicit attribute. noreferrer additionally suppresses the Referer header for cases where the destination should not see where the click came from."
	}
}

func attrLabelFor(tag string) string {
	if tag == "form" {
		return "action"
	}
	return "href"
}

File: internal/checks/target_blank_noopener.lua
-- target-blank-noopener: flags <a>, <area>, <form> with
-- target="_blank" that omit rel="noopener" / "noreferrer". The new
-- browsing context receives a live window.opener handle - reverse-
-- tabnabbing surface. Severity climbs to Medium for cross-origin
-- destinations.

local check = {
  name  = "target-blank-noopener",
  level = "passive",
  scope = "page",
  cwe   = "CWE-1022",
  owasp = "A05:2021 Security Misconfiguration",
  tier  = "passive",
}

local INTERESTING_TAGS = { "base", "a", "area", "form" }

-- rel is space-separated; any of noopener / noreferrer counts as safe.
local function rel_has_safe_token(rel)
  for tok in string.gmatch(rel, "%S+") do
    local lo = string.lower(tok)
    if lo == "noopener" or lo == "noreferrer" then
      return true
    end
  end
  return false
end

local function attr_label_for(tag)
  if tag == "form" then return "action" end
  return "href"
end

local function href_attr_for(tag)
  if tag == "form" then return "action" end
  return "href"
end

local function build_title(tag, cross_origin)
  local origin = "same-origin"
  if cross_origin then origin = "cross-origin" end
  return string.format('<%s target="_blank"> to %s URL without rel="noopener"', tag, origin)
end

local function build_detail(page_url, tag, raw, resolved, cross_origin)
  local detail = string.format(
    'Page %s contains <%s target="_blank" %s="%s"> resolving to %s without rel="noopener" or rel="noreferrer". ',
    page_url, tag, attr_label_for(tag), raw, resolved)
  detail = detail .. "The new browsing context receives a live window.opener handle to this page; "
  detail = detail .. 'script in the destination can call window.opener.location = "..." to silently navigate this tab to a phishing page (reverse tabnabbing). '
  if cross_origin then
    detail = detail .. "The destination is cross-origin, so any compromise or hostile content on that origin can pivot back into this site's tab."
  else
    detail = detail .. "The destination is same-origin, so direct impact is limited, but the missing attribute is still defense-in-depth worth fixing."
  end
  return detail
end

local function build_remediation(tag)
  if tag == "form" then
    return 'Add rel="noopener noreferrer" to the <form> element. Forms with target="_blank" did not get the same browser-default noopener treatment that anchors received, so the explicit attribute is the only portable guarantee.'
  end
  return 'Add rel="noopener noreferrer" to the element (e.g. <a href="..." target="_blank" rel="noopener noreferrer">). Modern browsers default anchors with target="_blank" to noopener, but older browsers, embedded webviews, and any code that opens windows via JavaScript still rely on the explicit attribute. noreferrer additionally suppresses the Referer header for cases where the destination should not see where the click came from.'
end

function check.run(ctx)
  local snap, err = ctx:ensure_response()
  if err then return nil, err end
  if not ctx.body.is_html_ct(snap.headers:get("Content-Type")) then
    return nil
  end
  if snap.body == "" then return nil end

  local page_url, perr = ctx.url.parse(ctx.page.url)
  if perr or not page_url or page_url.host == "" then return nil end

  local tags = ctx.html.iter_tags(snap.body, INTERESTING_TAGS)
  local base_url = ctx.page.url

  local evidence = ctx.evidence.build {
    method  = "GET",
    url     = ctx.page.url,
    status  = snap.status,
    headers = snap.headers,
  }

  local findings = {}
  local seen = {}

  for _, tag in ipairs(tags) do
    if tag.tag == "base" then
      local href = tag.attr["href"]
      if href and href ~= "" then
        local resolved, ok = ctx.html.resolve_ref(base_url, href)
        if ok then base_url = resolved end
      end
    else
      local target = string.lower((tag.attr["target"] or ""):gsub("^%s+", ""):gsub("%s+$", ""))
      if target == "_blank" then
        local rel = tag.attr["rel"] or ""
        if not rel_has_safe_token(rel) then
          local href = tag.attr[href_attr_for(tag.tag)] or ""
          local resolved, ok = ctx.html.resolve_ref(base_url, href)
          if ok then
            local ru, urerr = ctx.url.parse(resolved)
            if not urerr and ru and ru.host ~= ""
                and (ru.scheme == "http" or ru.scheme == "https") then
              local key = "ref:" .. tag.tag .. "|" .. resolved
              if not seen[key] then
                seen[key] = true
                local cross_origin = string.lower(ru.hostname) ~= string.lower(page_url.hostname)
                local severity = "low"
                if cross_origin then severity = "medium" end
                findings[#findings + 1] = {
                  severity    = ctx.severity[severity],
                  title       = build_title(tag.tag, cross_origin),
                  detail      = build_detail(ctx.page.url, tag.tag, href, resolved, cross_origin),
                  remediation = build_remediation(tag.tag),
                  evidence    = evidence,
                  dedupe_parts = { key },
                }
              end
            end
          end
        end
      end
    end
  end

  -- Stable order so per-page reports diff cleanly across runs.
  table.sort(findings, function(a, b)
    return a.dedupe_parts[1] < b.dedupe_parts[1]
  end)

  return findings
end

return check
